CollectionType should be an enum class

rniwa opened the pull request:
#### 8efe93387102c06cb10d329f5bf6d01abccdb5b0
<pre>
CollectionType should be an enum class
<a href="https://bugs.webkit.org/show_bug.cgi?id=256504">https://bugs.webkit.org/show_bug.cgi?id=256504</a>

Reviewed by Chris Dumez.

* Source/WebCore/bindings/js/JSHTMLCollectionCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/dom/AllDescendantsCollection.h:
(WebCore::AllDescendantsCollection::create):
* Source/WebCore/dom/ClassCollection.cpp:
(WebCore::ClassCollection::create):
* Source/WebCore/dom/ClassCollection.h:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::getElementsByTagName):
(WebCore::ContainerNode::getElementsByClassName):
(WebCore::ContainerNode::children):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::images):
(WebCore::Document::applets):
(WebCore::Document::embeds):
(WebCore::Document::plugins):
(WebCore::Document::scripts):
(WebCore::Document::links):
(WebCore::Document::forms):
(WebCore::Document::anchors):
(WebCore::Document::all):
(WebCore::Document::allFilteredByName):
(WebCore::Document::windowNamedItems):
(WebCore::Document::documentNamedItems):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/NodeRareData.h:
(WebCore::NodeListsNodeData::namedCollectionKey):
* Source/WebCore/dom/TagCollection.cpp:
(WebCore::TagCollectionNS::TagCollectionNS):
(WebCore::TagCollection::TagCollection):
(WebCore::HTMLTagCollection::HTMLTagCollection):
* Source/WebCore/dom/TagCollection.h:
* Source/WebCore/html/CachedHTMLCollectionInlines.h:
(WebCore::traversalType&gt;::namedItem const):
* Source/WebCore/html/CollectionType.h:
(): Deleted.
* Source/WebCore/html/GenericCachedHTMLCollection.cpp:
(WebCore::GenericCachedHTMLCollection&lt;traversalType&gt;::elementMatches const):
* Source/WebCore/html/HTMLAllCollection.cpp:
(WebCore::HTMLAllNamedSubCollection::HTMLAllNamedSubCollection):
* Source/WebCore/html/HTMLAllCollection.h:
* Source/WebCore/html/HTMLCollection.cpp:
(WebCore::HTMLCollection::rootTypeFromCollectionType):
(WebCore::invalidationTypeExcludingIdAndNameAttributes):
(WebCore::HTMLCollection::HTMLCollection):
(WebCore::HTMLCollection::~HTMLCollection):
(WebCore::HTMLCollection::updateNamedElementCache const):
* Source/WebCore/html/HTMLDataListElement.cpp:
(WebCore::HTMLDataListElement::options):
* Source/WebCore/html/HTMLFieldSetElement.cpp:
(WebCore::HTMLFieldSetElement::elements):
* Source/WebCore/html/HTMLFormControlsCollection.cpp:
(WebCore::HTMLFormControlsCollection::HTMLFormControlsCollection):
* Source/WebCore/html/HTMLFormControlsCollection.h:
* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::elements):
* Source/WebCore/html/HTMLMapElement.cpp:
(WebCore::HTMLMapElement::areas):
* Source/WebCore/html/HTMLNameCollection.h:
* Source/WebCore/html/HTMLNameCollectionInlines.h:
(WebCore::traversalType&gt;::~HTMLNameCollection):
* Source/WebCore/html/HTMLOptionsCollection.cpp:
(WebCore::HTMLOptionsCollection::HTMLOptionsCollection):
* Source/WebCore/html/HTMLOptionsCollection.h:
* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::selectedOptions):
(WebCore::HTMLSelectElement::options):
(WebCore::HTMLSelectElement::invalidateSelectedItems):
(WebCore::HTMLSelectElement::setRecalcListItems):
* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::rows):
(WebCore::HTMLTableElement::tBodies):
* Source/WebCore/html/HTMLTableRowElement.cpp:
(WebCore::HTMLTableRowElement::cells):
* Source/WebCore/html/HTMLTableRowsCollection.cpp:
(WebCore::HTMLTableRowsCollection::HTMLTableRowsCollection):
(WebCore::HTMLTableRowsCollection::create):
* Source/WebCore/html/HTMLTableRowsCollection.h:
* Source/WebCore/html/HTMLTableSectionElement.cpp:
(WebCore::HTMLTableSectionElement::rows):
* Source/WebKitLegacy/mac/DOM/DOMHTML.mm:
(kitClass):

Canonical link: <a href="https://commits.webkit.org/263894@main">https://commits.webkit.org/263894@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/18b765cc46a7220bf00aeb5d329910d9abd2940d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5903 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6076 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6263 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7455 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6266 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5900 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6295 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6037 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/8678 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6011 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6049 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5346 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7513 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3526 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5322 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13261 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5391 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5400 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7610 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5849 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4794 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5282 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9406 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/705 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5643 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/NodeRareData.h:
```diff
@@ -188,9 +188,9 @@ class NodeListsNodeData {
+    std::pair<uint8_t, AtomString> namedCollectionKey(CollectionType type, const AtomString& name)
     {
-        return std::pair<unsigned char, AtomString>(type, name);
+        return std::pair<uint8_t, AtomString>(static_cast<uint8_t>(type), name);
```
Might be nice to use std::underlying_type

rniwa replied:
Sure, we can do that.

cdumez reviewed (approved):
LGTM

rniwa committed:
CollectionType should be an enum class
https://bugs.webkit.org/show_bug.cgi?id=256504

Reviewed by Chris Dumez.

* Source/WebCore/bindings/js/JSHTMLCollectionCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/dom/AllDescendantsCollection.h:
(WebCore::AllDescendantsCollection::create):
* Source/WebCore/dom/ClassCollection.cpp:
(WebCore::ClassCollection::create):
* Source/WebCore/dom/ClassCollection.h:
* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::getElementsByTagName):
(WebCore::ContainerNode::getElementsByClassName):
(WebCore::ContainerNode::children):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::images):
(WebCore::Document::applets):
(WebCore::Document::embeds):
(WebCore::Document::plugins):
(WebCore::Document::scripts):
(WebCore::Document::links):
(WebCore::Document::forms):
(WebCore::Document::anchors):
(WebCore::Document::all):
(WebCore::Document::allFilteredByName):
(WebCore::Document::windowNamedItems):
(WebCore::Document::documentNamedItems):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/NodeRareData.h:
(WebCore::NodeListsNodeData::namedCollectionKey):
* Source/WebCore/dom/TagCollection.cpp:
(WebCore::TagCollectionNS::TagCollectionNS):
(WebCore::TagCollection::TagCollection):
(WebCore::HTMLTagCollection::HTMLTagCollection):
* Source/WebCore/dom/TagCollection.h:
* Source/WebCore/html/CachedHTMLCollectionInlines.h:
(WebCore::traversalType>::namedItem const):
* Source/WebCore/html/CollectionType.h:
(): Deleted.
* Source/WebCore/html/GenericCachedHTMLCollection.cpp:
(WebCore::GenericCachedHTMLCollection<traversalType>::elementMatches const):
* Source/WebCore/html/HTMLAllCollection.cpp:
(WebCore::HTMLAllNamedSubCollection::HTMLAllNamedSubCollection):
* Source/WebCore/html/HTMLAllCollection.h:
* Source/WebCore/html/HTMLCollection.cpp:
(WebCore::HTMLCollection::rootTypeFromCollectionType):
(WebCore::invalidationTypeExcludingIdAndNameAttributes):
(WebCore::HTMLCollection::HTMLCollection):
(WebCore::HTMLCollection::~HTMLCollection):
(WebCore::HTMLCollection::updateNamedElementCache const):
* Source/WebCore/html/HTMLDataListElement.cpp:
(WebCore::HTMLDataListElement::options):
* Source/WebCore/html/HTMLFieldSetElement.cpp:
(WebCore::HTMLFieldSetElement::elements):
* Source/WebCore/html/HTMLFormControlsCollection.cpp:
(WebCore::HTMLFormControlsCollection::HTMLFormControlsCollection):
* Source/WebCore/html/HTMLFormControlsCollection.h:
* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::elements):
* Source/WebCore/html/HTMLMapElement.cpp:
(WebCore::HTMLMapElement::areas):
* Source/WebCore/html/HTMLNameCollection.h:
* Source/WebCore/html/HTMLNameCollectionInlines.h:
(WebCore::traversalType>::~HTMLNameCollection):
* Source/WebCore/html/HTMLOptionsCollection.cpp:
(WebCore::HTMLOptionsCollection::HTMLOptionsCollection):
* Source/WebCore/html/HTMLOptionsCollection.h:
* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::selectedOptions):
(WebCore::HTMLSelectElement::options):
(WebCore::HTMLSelectElement::invalidateSelectedItems):
(WebCore::HTMLSelectElement::setRecalcListItems):
* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::rows):
(WebCore::HTMLTableElement::tBodies):
* Source/WebCore/html/HTMLTableRowElement.cpp:
(WebCore::HTMLTableRowElement::cells):
* Source/WebCore/html/HTMLTableRowsCollection.cpp:
(WebCore::HTMLTableRowsCollection::HTMLTableRowsCollection):
(WebCore::HTMLTableRowsCollection::create):
* Source/WebCore/html/HTMLTableRowsCollection.h:
* Source/WebCore/html/HTMLTableSectionElement.cpp:
(WebCore::HTMLTableSectionElement::rows):
* Source/WebKitLegacy/mac/DOM/DOMHTML.mm:
(kitClass):

Canonical link: https://commits.webkit.org/263894@main
